feat: get official field doc

golgoth31 opened the pull request:
Closes #456

## 📑 Description
Make an http call to the oficial k8S api reference page to get field documentation

## ✅ Checks
- [x] My pull request adheres to the code style of this project
- [x] My code requires changes to the documentation
- [x] I have updated the documentation as required
- [x] All the tests have passed

## ℹ Additional Information
This is currently a first preview to be discussed.

golgoth31 commented:
a simple exemple to show what it could looks like
before (from version v0.3.4):
![image](https://github.com/k8sgpt-ai/k8sgpt/assets/5741421/cb0a9e9a-b3ba-4470-b493-61f7cacc635a)

after (from main):
![image](https://github.com/k8sgpt-ai/k8sgpt/assets/5741421/2d7d4860-be50-4535-8790-0574a168032c)

matthisholleville commented:
Thank you for your contribution. I think it is an important and interesting information to display to the user. What do you think @AlexsJones ?

Can you test to deploy the following object and check that in case an object returns several errors, the information remains digestible please?

```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: fake-ingress
spec:
  rules:
  - host: fake-ingress.example.com
    http:
      paths:
      - backend:
          service:
            name: fake-service
            port:
              number: 80
        path: /
        pathType: Prefix
  tls:
  - hosts:
    - fake-ingress.example.com
    secretName: fake-secret
```

You should have this result with actual version :

```bash
0 k8sgpt/fake-ingress(fake-ingress)
- Error: Ingress k8sgpt/fake-ingress does not specify an Ingress class.
- Error: Ingress uses the service k8sgpt/fake-service which does not exist.
- Error: Ingress uses the secret k8sgpt/fake-secret as a TLS certificate which does not exist.
```

golgoth31 commented:
Hi,
with the actual version, I have:
0 default/fake-ingress(fake-ingress)
- Error: Ingress default/fake-ingress does not specify an Ingress class.
- Error: Ingress uses the service default/fake-service which does not exist.
- Error: Ingress uses the secret default/fake-secret as a TLS certificate which does not exist.
Error: Ingress default/fake-ingress doesn't specify an Ingress class. Ingress uses the service default/fake-service, which doesn't exist. Ingress uses the secret default/fake-secret as a TLS certificate, which doesn't exist.

Solution: 
1. Create the fake-service & fake-secret objects.
2. Add annotation "kubernetes.io/ingress.class: [class-name]" to fake-ingress definition.
3. Apply the changes to the cluster.

with the patch, I have:
0 default/fake-ingress(fake-ingress)
- Error: Ingress default/fake-ingress does not specify an Ingress class.
  Official Doc: ingressClassName is the name of an IngressClass cluster resource. Ingress controller implementations use this field to know whether they should be serving this Ingress resource, by a transitive connection (controller -> IngressClass -> Ingress resource). Although the `kubernetes.io/ingress.class` annotation (simple constant name) was never formally defined, it was widely supported by Ingress controllers to create a direct binding between Ingress controller and Ingress resources. Newly created Ingress resources should prefer using the field. However, even though the annotation is officially deprecated, for backwards compatibility reasons, ingress controllers should still honor that annotation if present.
- Error: Ingress uses the service default/fake-service which does not exist.
- Error: Ingress uses the secret default/fake-secret as a TLS certificate which does not exist.
Error: Ingress default/fake-ingress does not specify an Ingress class.
Solution: 
1. Define an Ingress class using the ingressClassName field in your YAML file.
2. If needed for backwards compatibility, add the kubernetes.io/ingress.class annotation.
3. Make sure the Ingress is referencing a valid service and secret.

matthisholleville commented:
Thank you. Shoudn't we have `Official Doc` part for each errors ?

golgoth31 commented:
I only extract the higher level field. Il have to find à way to parse also the subfields from the doc

golgoth31 commented:
I have an other way to find the information, I'll made some tests ... :)

golgoth31 commented:
Hi,
I have modified the patch to get the schema from the current server and extract data from it. The result is like this:
```
1 default/fake-ingress(fake-ingress)
- Error: Ingress default/fake-ingress does not specify an Ingress class.
  Kubernetes Doc: ingressClassName is the name of an IngressClass cluster resource. Ingress controller implementations use this field to know whether they should be serving this Ingress resource, by a transitive connection (controller -> IngressClass -> Ingress resource). Although the `kubernetes.io/ingress.class` annotation (simple constant name) was never formally defined, it was widely supported by Ingress controllers to create a direct binding between Ingress controller and Ingress resources. Newly created Ingress resources should prefer using the field. However, even though the annotation is officially deprecated, for backwards compatibility reasons, ingress controllers should still honor that annotation if present.
- Error: Ingress uses the service default/fake-service which does not exist.
  Kubernetes Doc: service references a service as a backend. This is a mutually exclusive setting with "Resource".
- Error: Ingress uses the secret default/fake-secret as a TLS certificate which does not exist.
  Kubernetes Doc: secretName is the name of the secret used to terminate TLS traffic on port 443. Field is left optional to allow TLS routing based on SNI hostname alone. If the SNI host in a listener conflicts with the "Host" header field used by an IngressRule, the SNI host is used for termination and value of the "Host" header is used for routing.
```

I've added a new field in the common.Failure struct to store the kubernetes doc string.
Once the kind, group and discovery client are set in the kubernetes.K8sApiReference struct, any field can be retreived by getting the path (eg: "spec.rules.http.paths.backend.service")

matthisholleville commented:
Thank you very much @golgoth31 ! We really appreciate the contribution and we could use it as option with `--with-doc` flag on analysis. Can you please add flag to use this option ?

golgoth31 commented:
> Thank you very much @golgoth31 ! We really appreciate the contribution and we could use it as option with `--with-doc` flag on analysis. Can you please add flag to use this option ?

Thank you,
I've added the with-doc flag, to ease the use of the flag, I move the getting of the openapi schema in the runanalisys function. 
I enforced the with-doc to false in server mode. I can make a PR to the protobuf if needed.

matthisholleville commented on pkg/analysis/output.go:
```diff
@@ -78,6 +78,9 @@ func (a *Analysis) textOutput() ([]byte, error) {
 			color.YellowString(result.Name), color.CyanString(result.ParentObject)))
 		for _, err := range result.Error {
 			output.WriteString(fmt.Sprintf("- %s %s\n", color.RedString("Error:"), color.RedString(err.Text)))
+			if err.KubernetesDoc != "" {
```
Shouldn't we add this on `json` output too ?

golgoth31 replied:
The kubernetesDoc is a filed of the failure struct, and so, part of the common.Result struct. At the end it's already in the json output.
```json
{
  "provider": "openai",
  "errors": null,
  "status": "ProblemDetected",
  "problems": 4,
  "results": [
    {
      "kind": "Ingress",
      "name": "default/fake-ingress",
      "error": [
        {
          "Text": "Ingress default/fake-ingress does not specify an Ingress class.",
          "KubernetesDoc": "ingressClassName is the name of an IngressClass cluster resource. Ingress controller implementations use this field to know whether they should be serving this Ingress resource, by a transitive connection (controller -\u003e IngressClass -\u003e Ingress resource). Although the `kubernetes.io/ingress.class` annotation (simple constant name) was never formally defined, it was widely supported by Ingress controllers to create a direct binding between Ingress controller and Ingress resources. Newly created Ingress resources should prefer using the field. However, even though the annotation is officially deprecated, for backwards compatibility reasons, ingress controllers should still honor that annotation if present.",
          ...
```

matthisholleville commented on pkg/kubernetes/apireference.go:
```diff
@@ -0,0 +1,77 @@
+	paths := strings.Split(field, ".")
+	group := strings.Split(k.ApiVersion.Group, ".")
+
+	// openapiSchema := openapi_v2.Document{}
```
Shouldn't we remove this ?

golgoth31 replied:
yes, of course

matthisholleville commented on pkg/kubernetes/apireference.go:
```diff
@@ -0,0 +1,77 @@
+
+	// extract the startpoint by searching the highest leaf corresponding to the requested group qnd kind
+	for _, prop := range definitions {
+		if strings.HasSuffix(prop.GetName(), group[0]+"."+k.ApiVersion.Version+"."+k.Kind) {
```
Can we use `fmt.Sprintf` to be more consistent ?

golgoth31 replied:
fixed

matthisholleville reviewed (changes requested):
Left few comments. Good job ! Can you please add usage example in README please ?

matthisholleville reviewed (approved):
LGTM.

Another review @AlexsJones ?

golgoth31 commented:
The readme has been updated

AlexsJones commented:
This is a great first-time contribution thanks @golgoth31
